driver-core/src/main/java/com/datastax/driver/core/SchemaParser.java

```diff
@@ -63,6 +63,9 @@ void refresh(Cluster cluster,
                 assert rows.keyspaces != null;
                 Map<String, KeyspaceMetadata> keyspaces = buildKeyspaces(rows, cassandraVersion, cluster);
                 updateKeyspaces(metadata, metadata.keyspaces, keyspaces, targetKeyspace);
+                // If we rebuild all from scratch or have an updated keyspace, rebuild the token map
+                // since some replication on some keyspace may have changed
+                metadata.rebuildTokenMap();
```

Comment:
> I am still a bit worried bout this call being synchronous, but not atomic

I agree, but fixing that is much more involved: I think the best way would be to make `Metadata` an immutable object that contains a snapshot of *all* metadata at a given point in time. After a single call to `cluster.getMetadata()` the user would have the guarantee that all fields are consistent with each other (which is not possible currently since you call methods on a mutable object).
Internally we would handle updates by storing this object in a volatile field, and switching it with a new copy when something gets updated.
This is likely to break the API though.